### DISCUSSION TITLE
Template for making own driver for Tasmota

### DISCUSSION THREAD

**Discussion by phpcoder** [arendst/Tasmota#14205] [General]
Many people ask how to easily make your own integration for Tasmota firmware. I made a repository with some [templates and instructions](https://github.com/phpcoder/tasmota-custom-integration) how to start creating your own Tasmota driver. Hope it helps.

**Comment by arendst** [comment]
Thx.

Looking over the Interface code I prefer something like this:

```
bool Xdrv86(uint8_t function) {
  bool result = false;

  if (FUNC_INIT == function) {
    SSPMInit();
  }
  else if (Sspm) {
    switch (function) {
      case FUNC_LOOP:
        if (SspmSerial) { SSPMSerialInput(); }
        break;
      case FUNC_EVERY_100_MSECOND:
        SSPMEvery100ms();
        break;
      case FUNC_SET_DEVICE_POWER:
        result = SSPMSetDevicePower();
        break;
      case FUNC_JSON_APPEND:
        SSPMEnergyShow(true);
        break;
#ifdef USE_WEBSERVER
      case FUNC_WEB_SENSOR:
        SSPMEnergyShow(false);
        break;
#endif  // USE_WEBSERVER
      case FUNC_COMMAND:
        result = DecodeCommand(kSSPMCommands, SSPMCommand);
        break;
      case FUNC_BUTTON_PRESSED:
        result = SSPMButton();
        break;
    }
  }
  return result;
}
```

Notice the initial ``FUNC_INIT`` at the top. This should verify all parameters are valid for this driver to handle requests. If not it makes sure all other functions are not supported.

When the driver cannot perform it's services this allows to bail out out early for all other ``FUNC_`` calls speeding up processing AND makes cleaner code as you do not need to test every function for being allowed.

See driver ``xdrv_86_esp32_sonoff_spm.ino`` as an example.

**Comment by barbudor** [comment]
Another comment: in your 2nd driver you are defining large variables
https://github.com/phpcoder/tasmota-custom-integration/blob/48e3bc31128ceb95794101d0ef69286693649e3f/tasmota/xdrv_101_my_project_with_commands.ino#L45-L46

This would use RAM everytime the driver is compiled-in even if not used. This could be problematic for builds like mega or platinum that are pushing the limits of the Flash.
It a good practice to dynamically allocate RAM for your driver only if is in use.
Theo's SPM or my [HM330x](https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_93_hm330x.ino) drivers illustrate this

**Comment by phpcoder** [comment]
Thank you guys for comments, and special thanks to Theo for your great work.
I will modify my examples, however, I wanted offer something super simple for a quick start, not making a production grade code...

**Comment by nikito7** [comment]
How to pass parameters in a custom command?

```cpp

const char HanCommands[] PROGMEM =
    "|"  // No Prefix
    "HanDelay|"
    "HanHelp";

void (*const HanCommand[])(void) PROGMEM = {&CmdHanDelay,
                                            &CmdHanHelp};

void CmdHanHelp(void) {
  AddLog(LOG_LEVEL_INFO, PSTR("HanHelp: Hello!"));
  ResponseCmndDone();
}

void CmdHanDelay(void) {
  AddLog(
      LOG_LEVEL_INFO,
      PSTR(
          "HanDelay: Set modbus pause delay in seconds"));
  hanDelay = 300000;
  sprintf(hStatus, "Cmd");
  ResponseCmndDone();
}
```

I need to change hanDelay

**Reply by sfromis** [reply]
Here's an example of a command taking a number as argument:
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_support/support_command.ino#L1346-L1353
```
void CmndBlinkcount(void)
{
  if ((XdrvMailbox.payload >= 0) && (XdrvMailbox.payload <= 32000)) {
    Settings->blinkcount = XdrvMailbox.payload;  // 0 - 65535
    if (TasmotaGlobal.blink_counter) { TasmotaGlobal.blink_counter = Settings->blinkcount *2; }
  }
  ResponseCmndNumber(Settings->blinkcount);
}
```
